clippy_lints/src/unnecessary_indexing.rs

```diff
@@ -0,0 +1,208 @@
+                                .filter_map(|x| {
+                                    if let ExprKind::Let(l) = x.kind {
+                                        Some((l.init.span, snippet(cx, first_local.pat.span, "..").to_string()))
+                                    } else {
+                                        None
+                                    }
+                                })
+                                .collect::<Vec<_>>();
+
+                            if !extra_local_suggestions.is_empty() {
+                                diag.multipart_suggestion(
+                                    "initialize this variable to be the `Some` variant (may need dereferencing)",
+                                    extra_local_suggestions,
+                                    Applicability::Unspecified,
+                                );
+                            }
+                        }
+                        if !r.extra_exprs.is_empty() {
+                            let index_accesses = r
+                                .extra_exprs
+                                .iter()
+                                .map(|x| (x.span, snippet(cx, first_local.pat.span, "..").to_string()))
+                                .collect::<Vec<_>>();
+
+                            diag.multipart_suggestion(
+                                "set this index to be the `Some` variant (may need dereferencing)",
+                                index_accesses,
+                                Applicability::Unspecified,
+                            );
+                        }
+                    } else {
+                        let mut index_accesses = vec![(
+                            if_expr.cond.span,
+                            format!("let Some(element) = {}.first()", snippet(cx, receiver.span, "..")),
+                        )];
+                        index_accesses.extend(r.extra_exprs.iter().map(|x| (x.span, "element".to_owned())));
+
+                        diag.multipart_suggestion(
+                            "consider using if..let syntax (variable may need to be dereferenced)",
+                            index_accesses,
+                            Applicability::Unspecified,
+                        );
+                    }
+                },
+            );
+        }
+    }
+}
+
+struct IndexCheckResult<'a> {
+    // the receiver for the index operation, only Some in the event the indexing is via a direct primitive
+    index_receiver: Option<&'a Expr<'a>>,
+    // first local in the block - used as pattern for `Some(pat)`
+    first_local: Option<&'a LetStmt<'a>>,
+    // any other index expressions to replace with `pat` (or "element" if no local exists)
+    extra_exprs: Vec<&'a Expr<'a>>,
+}
+impl<'a> IndexCheckResult<'a> {
+    pub fn new() -> Self {
+        IndexCheckResult {
+            index_receiver: None,
+            first_local: None,
+            extra_exprs: vec![],
+        }
+    }
+}
```

Comment:
Unused

Reply:
Not quite sure how I missed this, oops

Reply:
I was surprised it didn't warn, I guess the `pub` throws it off